From d57324ac7324e1e7810b8601466572646f0782bd Mon Sep 17 00:00:00 2001
From: Philip Chimento <philip.chimento@gmail.com>
Date: Tue, 7 Jan 2020 23:52:00 -0500
Subject: [PATCH] build: Prefer signal.h to sys/signal.h

It seems that signal.h is the correct header to include, but on win32 it
is still sys/signal.h.

Closes: #295
---
 gjs/context.h              | 6 ++++--
 gjs/profiler.cpp           | 8 +++++---
 meson.build                | 5 ++++-
 tools/gjs-private-iwyu.imp | 4 ++--
 4 files changed, 15 insertions(+), 8 deletions(-)

diff --git a/gjs/context.h b/gjs/context.h
index b173c32c..dd6a36f7 100644
--- a/gjs/context.h
+++ b/gjs/context.h
@@ -30,8 +30,10 @@
 
 #include <stdbool.h>    /* IWYU pragma: keep */
 
-#ifndef _WIN32
-#    include <sys/signal.h> /* for siginfo_t */
+#ifdef HAVE_SIGNAL_H
+#    include <signal.h> /* for siginfo_t */
+#else
+#    include <sys/signal.h>
 #endif
 
 #include <glib-object.h>
diff --git a/gjs/profiler.cpp b/gjs/profiler.cpp
index 74c99f07..c4abaca8 100644
--- a/gjs/profiler.cpp
+++ b/gjs/profiler.cpp
@@ -21,10 +21,12 @@
  * IN THE SOFTWARE.
  */
 
-#include <config.h>  // for ENABLE_PROFILER, HAVE_SYS_SYSCALL_H, HAVE_UNISTD_H, HAVE_SYS_SIGNAL_H
+#include <config.h>  // for ENABLE_PROFILER, HAVE_SYS_SYSCALL_H, HAVE_UNISTD_H
 
-#ifdef HAVE_SYS_SIGNAL_H
-#	include <sys/signal.h>  // for siginfo_t, sigevent, ...
+#ifdef HAVE_SIGNAL_H
+#    include <signal.h>  // for siginfo_t, sigevent, sigaction, SIGPROF, ...
+#else
+#    include <sys/signal.h>
 #endif
 
 #include <glib-object.h>
diff --git a/meson.build b/meson.build
index 072a3e1d..2c188286 100644
--- a/meson.build
+++ b/meson.build
@@ -283,7 +283,10 @@ if build_readline
 endif
 header_conf.set('HAVE_SYS_SYSCALL_H', cxx.check_header('sys/syscall.h'))
 header_conf.set('HAVE_UNISTD_H', cxx.check_header('unistd.h'))
-header_conf.set('HAVE_SYS_SIGNAL_H', cxx.check_header('sys/signal.h'))
+have_signal_h = cxx.check_header('signal.h')
+header_conf.set('HAVE_SIGNAL_H', have_signal_h)
+header_conf.set('HAVE_SYS_SIGNAL_H', cxx.check_header('sys/signal.h',
+    required: not have_signal_h))
 
 # enable GNU extensions on systems that have them
 header_conf.set('_GNU_SOURCE', 1)
diff --git a/tools/gjs-private-iwyu.imp b/tools/gjs-private-iwyu.imp
index 43fe827c..19ffbcbd 100644
--- a/tools/gjs-private-iwyu.imp
+++ b/tools/gjs-private-iwyu.imp
@@ -3,8 +3,8 @@
   {"include": ["<bits/std_abs.h>", "private", "<stdlib.h>", "public"]},
   {"include": ["<bits/std_function.h>", "private", "<functional>", "public"]},
   {"include": ["@<bits/stdint-.*>", "private", "<stdint.h>", "public"]},
-  {"include": ["<bits/types/sigevent_t.h>", "private", "<sys/signal.h>", "public"]},
-  {"include": ["<bits/types/siginfo_t.h>", "private", "<sys/signal.h>", "public"]},
+  {"include": ["<bits/types/sigevent_t.h>", "private", "<signal.h>", "public"]},
+  {"include": ["<bits/types/siginfo_t.h>", "private", "<signal.h>", "public"]},
   {"include": ["<bits/types/struct_itimerspec.h>", "private", "<time.h>", "public"]},
   {"include": ["<bits/types/struct_timespec.h>", "private", "<time.h>", "public"]},
   {"include": ["<bits/types/timer_t.h>", "private", "<sys/types.h>", "public"]},
-- 
2.24.1

